docs/CodeTransformation.md

```diff
@@ -22,110 +22,154 @@ If you override the `transform` configuration option `babel-jest` will no longer
 You can write your own transformer. The API of a transformer is as follows:
 
 ```ts
-interface SyncTransformer<OptionType = unknown> {
+export interface ShouldInstrumentOptions
+  extends Pick<
+    Config.GlobalConfig,
+    | 'collectCoverage'
+    | 'collectCoverageFrom'
+    | 'collectCoverageOnlyFrom'
+    | 'coverageProvider'
+  > {
+  changedFiles?: Set<string>;
+  sourcesRelatedToTestsInChangedFiles?: Set<string>;
+}
+
+export interface Options
+  extends ShouldInstrumentOptions,
+    CallerTransformOptions {
+  isInternalModule?: boolean;
+}
+
+// This is fixed in source-map@0.7.x, but we can't upgrade yet since it's async
+interface FixedRawSourceMap extends Omit<RawSourceMap, 'version'> {
+  version: number;
+}
+
+// TODO: For Jest 26 normalize this (always structured data, never a string)
+export type TransformedSource =
+  | {code: string; map?: FixedRawSourceMap | string | null}
+  | string;
+
+export type TransformResult = TransformTypes.TransformResult;
+
+export interface CallerTransformOptions {
+  // names are copied from babel: https://babeljs.io/docs/en/options#caller
+  supportsDynamicImport: boolean;
+  supportsExportNamespaceFrom: boolean;
+  supportsStaticESM: boolean;
+  supportsTopLevelAwait: boolean;
+}
+
+export interface ReducedTransformOptions extends CallerTransformOptions {
+  instrument: boolean;
+}
+
+export interface RequireAndTranspileModuleOptions
+  extends ReducedTransformOptions {
+  applyInteropRequireDefault: boolean;
+}
+
+export type StringMap = Map<string, string>;
+
+export interface TransformOptions<OptionType = unknown>
+  extends ReducedTransformOptions {
+  /** a cached file system which is used in jest-runtime - useful to improve performance */
+  cacheFS: StringMap;
+  config: Config.ProjectConfig;
+  /** A stringified version of the configuration - useful in cache busting */
+  configString: string;
+  /** the options passed through Jest's config by the user */
+  transformerConfig: OptionType;
+}
```

Comment:
I agree it could be trimmed down a bit; that's why I trimmed it to begin with. Just misunderstood you and [thought you wanted it back](https://github.com/facebook/jest/pull/12407#discussion_r808136076) :) I can trim it down a bit. 

I guess the Transformer interfaces and the exports are the main things people are interested in, and then they can have a look at the code for details.

With regards to linking it back to the code, I am a bit unsure what to do/what you intended:

1.  [The comment I added yesterday](https://github.com/facebook/jest/pull/12407/files/2ac8acaa32c00d0f37ce6eac608854826a0be647#diff-878bcedd9971caed9cbdb29b52cc361e847292632ad339a0061a15a91b415420R24) in the Markdown would disappear when rendered on the homepage
1. I could add the same comment at the top of the interface definition
1. I could add an actual html link to the prose, but then that would almost never point to the correct interface definition due to versioning.

I think approach #2 is the least flaky approach.